### ISSUE TITLE
How to get >= 60 FPS output

### ISSUE THREAD

**Issue by akgaurav** [pedroSG94/rtmp-rtsp-stream-client-java#715]
Current code does not provide more than 30 to 35 FPS. The draw screen method takes more than 15~20 ms on an average followed by update frame more than 10 ms. 

How to achieve higher FPS with this code. Please help.

**Comment by pedroSG94** [comment]
Currently you can get more than 30fpsbusing camera1 or camera2 API. So this is not possible.
About drawScreen method and updateFrame, that time should depend of GPU used and if you are using lightopenglview or you reduce number of filters you can get better results

**Comment by akgaurav** [comment]
Hi pedroSG94,  Thanks for your reply. In camera2APi capturecallback I am able to get 60FPS output. But this reduces to 30 FPS in OnFrameCallback due to time comlexity of draw screen and update Frame. My requirement is full HD (1920 * 1080p) at 60 FPS
Also is it possible to receive the byteArray for some processing at 60FPS, and screen render only at 30 FPS ?

I am able to get 120 fps also in camera2 api capture callback using highSpeedCaptureSession

I believe camera is capturing at high FPS but screen rendering is slowing it down.

**Comment by pedroSG94** [comment]
> Hi pedroSG94, Thanks for your reply. In camera2APi capturecallback I am able to get 60FPS output. But this reduces to 30 FPS in OnFrameCallback due to time comlexity of draw screen and update Frame

Can you share a code example to check fps using camera2API?

> Also is it possible to receive the byteArray for some processing at 60FPS, and screen render only at 30 FPS ?

It is possible but library is developed to render preview and encoder to same fps. It could be interested if I can get a device that render to 60fps with camera2.


> I am able to get 120 fps also in camera2 api capture callback using highSpeedCaptureSession

This is used for slow motion so this is not recommended and I need render screen and encoder separated to work.

**Comment by akgaurav** [comment]
_It could be interested if I can get a device that render to 60fps with camera2._

I just write below code to get 60 fps output
builder.set(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE, new Range<>(60,60))

This sets the capture request FPS range to 60FPS.

Can you share a code example to check fps using camera2API?
-- To check FPS I just write log after receiving 60th frame and check the time in LOGCAT.
  basically I will print in Logcat for frame 60, 120, 180, 240 and so on.

I will try lightOpenGlView as suggested by you. Could you let me know the difference b/w lightOpenGLView and OpenGlView

**Comment by pedroSG94** [comment]
> I just write below code to get 60 fps output
> builder.set(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE, new Range<>(60,60))

I know this method but it is not possible with oneplus 7T. The limit is 30fps so I can't test this.
I did a calbback to check fps as you did but I only can get 3'fps from camera2 api.



> I will try lightOpenGlView as suggested by you. Could you let me know the difference b/w lightOpenGLView and OpenGlView

With LightOpenGLView you are using less GPU (but you can't use filters), so you should get better performance.

**Comment by akgaurav** [comment]
Hi pedroSG9,  I tried using LightOpenGlView as advised by you. Unfortunately I am still not able to get more than 30 ~35 FPS. 
Reducing the resolution to 240p results in  very minor improvement (taking FPS to 43). Any idea how can we optimize OpenGl code ? How to use hardware/GPU acceleration with OpenGL ?

**Comment by akgaurav** [comment]
> In my case using oneplus 7T with openglview and recording (no filters added). In my example app (resolution 480p) render 1 frame take around 5ms that means a total of 200fps so the limit in this case is camera api.
> 1080p take around 6-7ms -> 142fps

How did you measure this -  Does it take only 5 ms  for  (makeCurrent + updateFrame + drawscreen + swapbuffer ) ?

**Comment by pedroSG94** [comment]
> Reducing the resolution to 240p results in very minor improvement (taking FPS to 43). Any idea how can we optimize OpenGl code ? How to use hardware/GPU acceleration with OpenGL ?

OpenGL is hardware accelerated.
https://en.m.wikipedia.org/wiki/OpenGL#:~:text=OpenGL%20(Open%20Graphics%20Library)%20is,Silicon%20Graphics%2C%20Inc.
I have no access to Samsung S9 so I can't check where is the problem but I can try check ms time with Samsung S7.
For now, If you are using filters disable all filters to discard performance problems on it.

> How did you measure this - Does it take only 5 ms for (makeCurrent + updateFrame + drawscreen + swapbuffer ) ?

5ms from this line:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OpenGlView.java#L129
To this:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OpenGlView.java#L166
This is the loop where you render a frame

**Comment by pedroSG94** [comment]
With Samsung S7 in 1080p render a frame take 9-11ms -> 90-111fps
Can you check inside drawScreen method? Where do you need more ms to proccess?
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/ScreenRender.java
Which drawScreen take 20ms?
Preview:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OpenGlView.java#L133
Encoder:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OpenGlView.java#L144

**Comment by akgaurav** [comment]
> Go to device settings, developer options and check if you have disable MSAA for OpenGL. Also force GPU acceleration in that section.

I enabled MSAA and Force GPU acceleration. It did not improve the FPS. 

> Which drawScreen take 20ms?

Preview drawScreen takes 20ms. I might be using older version of your code, Hope that will not impact FPS.


I am attaching Log which shows time in ms for execution of rendering methods

2018-02-06 18:51:08.607 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.617 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 10
2018-02-06 18:51:08.629 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 11
2018-02-06 18:51:08.634 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 5
2018-02-06 18:51:08.635 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.638 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 3
2018-02-06 18:51:08.664 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 26
2018-02-06 18:51:08.667 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 3
2018-02-06 18:51:08.667 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.685 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 18
2018-02-06 18:51:08.695 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 10
2018-02-06 18:51:08.701 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 5
2018-02-06 18:51:08.702 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.703 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 1
2018-02-06 18:51:08.721 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 18
2018-02-06 18:51:08.725 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 4
2018-02-06 18:51:08.725 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.733 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 8
2018-02-06 18:51:08.758 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 25
2018-02-06 18:51:08.763 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 5
2018-02-06 18:51:08.763 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.765 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 2
2018-02-06 18:51:08.785 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 20
2018-02-06 18:51:08.788 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 3
2018-02-06 18:51:08.788 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.801 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 13
2018-02-06 18:51:08.810 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 9
2018-02-06 18:51:08.814 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 4
2018-02-06 18:51:08.814 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.818 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 4
2018-02-06 18:51:08.838 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 20
2018-02-06 18:51:08.842 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 4
2018-02-06 18:51:08.842 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.852 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 10
2018-02-06 18:51:08.874 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 22
2018-02-06 18:51:08.880 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 5
2018-02-06 18:51:08.880 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.884 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 4
2018-02-06 18:51:08.902 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 18
2018-02-06 18:51:08.904 19652-21283/com.myOrg.myApp I/System.out: after swapBuffer 2
2018-02-06 18:51:08.904 19652-21283/com.myOrg.myApp I/System.out: after makeCurrent 0
2018-02-06 18:51:08.920 19652-21283/com.myOrg.myApp I/System.out: after updateFrame 16
2018-02-06 18:51:08.930 19652-21283/com.myOrg.myApp I/System.out: after drawScreen 10

**Comment by pedroSG94** [comment]
I mean that time but inside this method:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/ScreenRender.java#L80
Also lets try using background mode to discard problems rendering device screen. Replace:
```java
RtmpCamera2(OpenGlView openGlView, ConnectCheckerRtmp connectChecker)
```
To:
```java
//useOpengl = true
RtmpCamera2(Context context, boolean useOpengl, ConnectCheckerRtmp connectChecker)
```

**Comment by akgaurav** [comment]
Hi pedroSG94, 

I found the solution to the FPS problem. I was viewing the device remotely using SCRCPY. This reduces the speed to 40 FPS.
When I directly start the app on device. I am getting 60 FPS output. Tested with LighOpenGlView, MSAA & Forced Acceleration enabled

**Comment by pedroSG94** [comment]
It good to know it.
I recommend you disable MSAA and disable force acceleration.
Closing issue, you can reopen it if you need
